### PR TITLE
feat(luadev): better vim api completion

### PR DESCRIPTION
**Pull request by meijieru**
# Description

Use [lua-dev.nvim](https://github.com/folke/lua-dev.nvim) instead of `cmp-nvim-lua`.
It provides better signature support.

Origin
image

Now
image

It could also have plugin's completion by modifying [here](https://github.com/LunarVim/LunarVim/pull/2043/files#diff-072b5a568c83434dda5a33ca67ecdcb46358ba94308236464a9dc8b1578bbff7R7)

### CODE REVIEW

File: lua/lvim/plugins.lua [hunk 1]
```diff
@@ -3,7 +3,6 @@ local commit = {
   cmp_buffer = "a706dc69c49110038fe570e5c9c33d6d4f67015b",
   cmp_luasnip = "16832bb50e760223a403ffa3042859845dd9ef9d",
   cmp_nvim_lsp = "134117299ff9e34adde30a735cd8ca9cf8f3db81",
```
Comment (kylo252) [inline lua/lvim/plugins.lua:5]:
```suggestion
  cmp_nvim_lsp = "134117299ff9e34adde30a735cd8ca9cf8f3db81",
  cmp_nvim_lua = "d276254e7198ab7d00f117e88e223b4bd8c02d21",
```

File: lua/lvim/lsp/providers/sumneko_lua.lua [hunk 2]
```diff
@@ -1,19 +1,27 @@
...
+          preloadFileSize = 10000,
         },
-        maxPreload = 100000,
-        preloadFileSize = 10000,
       },
     },
   },
 }
+
 return opts
```
Comment (kylo252) [inline lua/lvim/lsp/providers/sumneko_lua.lua:27]: I can't make suggestion over deleted lines 😄 
We should use a protected call here instead
```lua
local opts = {
  lspconfig = {
    settings = {
      Lua = {
        diagnostics = {
          globals = { "vim", "lvim" },
        },
        workspace = {
          library = {
            [require("lvim.utils").join_paths(get_runtime_dir(), "lvim", "lua")] = true,
          },
          maxPreload = 100000,
          preloadFileSize = 10000,
        },
      },
    },
  },
}

local lua_dev_loaded, lua_dev = pcall(require, "lua-dev")
if not lua_dev_loaded then
  return opts
end

local dev_opts = vim.tbl_deep_extend("force", {
  library = {
    vimruntime = true, -- runtime path
    types = true, -- full signature, docs and completion of vim.api, vim.treesitter, vim.lsp and others
    -- plugins = true, -- installed opt or start plugins in packpath
    -- you can also specify the list of plugins to make available as a workspace library
    plugins = { "plenary.nvim" },
  },
}, opts)

return lua_dev.setup(dev_opts)
```

File: lua/lvim/plugins.lua [hunk 3]
```diff
@@ -114,10 +113,7 @@ return {
     "hrsh7th/cmp-path",
     commit = commit.cmp_path,
   },
-  {
-    "hrsh7th/cmp-nvim-lua",
-    commit = commit.cmp_nvim_lua,
-  },
+  { "folke/lua-dev.nvim" },
```
Comment (kylo252) [inline lua/lvim/plugins.lua:116]:
```suggestion
  { "folke/lua-dev.nvim", commit = commit.lua_dev },
```

And add this at line 15

```lua
  lua_dev = "1933c7e014e69484572b7fa1bf73bc51c42f10f4",
```

**Review by kylo252 — CHANGES_REQUESTED** [review]

File: lua/lvim/plugins.lua (hunk 1 above)
Comment (meijieru) [reply lua/lvim/plugins.lua:5]: It seems OK to keep that, as we don't have tabnine but have it there too.

File: lua/lvim/lsp/providers/sumneko_lua.lua [hunk 4]
```diff
@@ -7,13 +7,27 @@ local opts = {
...
+
+local dev_opts = vim.tbl_deep_extend("force", {
+  library = {
+    vimruntime = true, -- runtime path
+    types = true, -- full signature, docs and completion of vim.api, vim.treesitter, vim.lsp and others
+    -- plugins = true, -- installed opt or start plugins in packpath
+    -- you can also specify the list of plugins to make available as a workspace library
+    plugins = { "plenary.nvim" },
+  },
+}, { lspconfig = opts })
```
Comment (kylo252) [inline lua/lvim/lsp/providers/sumneko_lua.lua:31]: Sorry! I forgot that we can pass the other opts like this, so there's no need for merging them.
```suggestion
local dev_opts =  {
  library = {
    vimruntime = true, -- runtime path
    types = true, -- full signature, docs and completion of vim.api, vim.treesitter, vim.lsp and others
    -- plugins = true, -- installed opt or start plugins in packpath
    -- you can also specify the list of plugins to make available as a workspace library
    plugins = { "plenary.nvim" },
  },
  lspconfig = opts
}
```

**Review by kylo252 — CHANGES_REQUESTED** [review]

File: lua/lvim/lsp/providers/sumneko_lua.lua (hunk 4 above)
Comment (meijieru) [reply lua/lvim/lsp/providers/sumneko_lua.lua]: done

**Comment by chase** [comment]
I use both together with no issues so far:
https://github.com/chase/lvim/blob/main/lua/lvim/lsp/providers/sumneko_lua.lua

**Comment by kylo252** [comment]
@chase, should we add more plugins to the library list?

**Review by kylo252 — COMMENTED** [review]
I just noticed that this breaks the nlsp-settings' integration :disappointed: 

Not sure what to do here, so I'll leave it to you guys to decide.
@chase, @abzcoding

**Comment by abzcoding** [comment]
> I just noticed that this breaks the nlsp-settings' integration 😞
> 
> Not sure what to do here, so I'll leave it to you guys to decide. @chase, @abzcoding

atm i'm using both with proper nlsp-settings support, so i'm fine the way lvim currently is (without this PR) 😅

**Comment by meijieru** [comment]
Any idea on why this happened?

**Comment by kylo252** [comment]
> Any idea on why this happened?

Do you mean why `nlsp-settings` stopped working? It's probably just certain flags that are always overwritten by lua-dev. Overall, it seems to be working for most settings for me. Please rebase this again so we can get it merged :)

**Comment by meijieru** [comment]
> > Any idea on why this happened?
> 
> Do you mean why `nlsp-settings` stopped working? It's probably just certain flags that are always overwritten by lua-dev. Overall, it seems to be working for most settings for me. Please rebase this again so we can get it merged :)

done.

**Review by kylo252 — APPROVED** [review]
LGTM 🚀 

---

For reference, I had to re-save `sumneko_lua.json` file to force nlsp-settings to refresh, so maybe we would have to trigger that `load()` manually.

**Comment by kylo252** [comment]
@meijieru, thanks!

**Pull request merged**